### PR TITLE
Split CI test groups into concurrent jobs

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
# Description

> [!Important]  
> This PR also contains the changes made as a part of #3279, so both PRs should be merged together, with that one being merged before this one.

Improves CI time by creating concurrent jobs for separate `nox` sessions for running tests, also adds the concurrency setting to cancel in-progress jobs.

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: .github/workflows/test_on_push.yml [hunk 1]
```diff
@@ -19,6 +19,8 @@ jobs:
           cancel_others: "true"
           paths_ignore: '["**/README.md"]'
 
+# TODO: add concurrency syntax for groups and cancelling in-progress jobs
+
```
Comment (Saransh-cpp) [inline .github/workflows/test_on_push.yml:23]: See https://github.com/scikit-hep/vector/blob/349aca6123873bcbaa1470d411304dab4009327a/.github/workflows/ci.yml#L11-L15 for inspiration

**Comment by agriyakhetarpal** [comment]
Failure for a job cancels other jobs in the matrix and the further jobs that depend on it, as intended: https://github.com/pybamm-team/PyBaMM/actions/runs/5823292074. Doctests on other platforms were cancelled because the ones on Windows and Python 3.11 failed, and subsequently the examples tests did not run.

File: .github/workflows/test_on_push.yml (hunk 1 above)
Comment (agriyakhetarpal) [reply .github/workflows/test_on_push.yml]: FIxed in https://github.com/pybamm-team/PyBaMM/pull/3264/commits/ddc0fc587bbf95000ac2d4344f83b80d5ff0f94f, I improved the `vector` one to work on pushes outside of PRs as well

**Comment by agriyakhetarpal** [comment]
I turned on all tests (unit, coverage, integration, doctests, examples) for all platforms temporarily. Just wanting to check runtimes for the test suite and those for individual jobs

**Comment by agriyakhetarpal** [comment]
Weird that the coverage tests on macOS are [failing](https://github.com/pybamm-team/PyBaMM/actions/runs/5824687678/job/15794395295) in the workflows because of installing `[odes]`. It is unable to create wheels for `scikits.odes` (which is a common error), but this session passes for me locally (it iooked to me like a problem with the `set_environment_variables` function in `noxfile.py`, but it wasn't—the `SUNDIALS_INST` and `LD_LIBRARY_PATH` environment variables are indeed being set correctly). @Saransh-cpp we could skip it like we were doing before, but in theory this shouldn't be the case: let me know

**Comment by Saransh-cpp** [comment]
> @Saransh-cpp we could skip it like we were doing before, but in theory this shouldn't be the case: let me know

Yes, we can skip them for now

**Comment by agriyakhetarpal** [comment]
> Yes, we can skip them for now

Even more weird that `scikits.odes` now fails on Ubuntu with Python 3.11 too, which definitely should not happen

**Review by Saransh-cpp — COMMENTED** [review]
We could also run integration tests in parallel to unit tests. This will save us a lot of CI time, and we can just let them fail if the unit tests are failing.

**Comment by agriyakhetarpal** [comment]
Done, although it might not save us CI time because the examples tests will take ~35-40 minutes anyway

File: .github/workflows/run_periodic_tests.yml [hunk 2]
```diff
@@ -102,7 +102,7 @@ jobs:
 
       - name: Upload coverage report
         if: matrix.os == 'ubuntu-latest' && matrix.python-version == 3.11
-        uses: codecov/codecov-action@v2.1.0
+        uses: codecov/codecov-action@v3.1.4
```
Comment (Saransh-cpp) [inline .github/workflows/run_periodic_tests.yml:105]: These versions go stale very often in PyBaMM. I have opened #3276 for this.

**Review by Saransh-cpp — APPROVED** [review]
Looks great, thanks @agriyakhetarpal! I will let this sit for a while for others to take a look at 🙂

File: .github/workflows/run_periodic_tests.yml (hunk 2 above)
Comment (agriyakhetarpal) [reply .github/workflows/run_periodic_tests.yml:105]: Awesome!

**Comment by agriyakhetarpal** [comment]
Also, do look at the document too that I had shared on Slack, when you are able to: it contains some more ideas that I can implement in this PR (or possibly in another PR that would get merged before this one does). Hopefully @tinosulzer has further suggestions on those points

**Comment by agriyakhetarpal** [comment]
The CI runtime has indeed halved from the earlier ~65 minutes—the new examples job ran in 34 minutes. This means we can get sub-30-minute times on a good day 🎉

**Comment by Saransh-cpp** [comment]
@agriyakhetarpal you can rebase this now!

**Comment by agriyakhetarpal** [comment]
I know that we need eyes from others on this, a note will need to merge this soon—otherwise the example scripts will not be tested on further PRs (see #3279). We were ideally supposed to keep both of them open and merge them one after the other

**Pull request merged**